### PR TITLE
Liveness loop nesting forest

### PR DESCRIPTION
**Pull request by clonker**
depends on https://github.com/ethereum/solidity/pull/15418 (merged)

### CODE REVIEW

File: libyul/backends/evm/SSACFGTopologicalSort.cpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+			m_potentialBackEdges.emplace_back(_vertex, _exitBlock.value);
+	});
+
+	m_postOrder.push_back(_vertex);
+}
+
+std::vector<size_t> const& ForwardSSACFGTopologicalSort::preOrder() const { return m_preOrder; }
+std::vector<size_t> const& ForwardSSACFGTopologicalSort::postOrder() const { return m_postOrder; }
+std::set<size_t> const& ForwardSSACFGTopologicalSort::backEdgeTargets() const { return m_backEdgeTargets; }
+SSACFG const& ForwardSSACFGTopologicalSort::cfg() const { return m_cfg; }
```
Comment (matheusaaguiar) [inline libyul/backends/evm/SSACFGTopologicalSort.cpp]: why not defined in the header?

**Review by matheusaaguiar — DISMISSED** [review]
Checked against the algorithm and everything looks ok!

File: libyul/backends/evm/SSACFGTopologicalSort.cpp (hunk 1 above)
Comment (clonker) [reply libyul/backends/evm/SSACFGTopologicalSort.cpp]: I didn't give it much thought at the time. So now, after some thinking, here are my considerations: 

- It would lead to more inlining opportunities for the compiler (if it makes a difference, stuff like is also picked up by LTO - do we have it enabled?). So if this was used in a tight loop then it might very well make a performance difference (which has to be profiled, doubt it would be a big one).
- On the other hand, doing it like this can bring down compile times and if something has to be changed in the implementation, incremental builds are faster. Although that seems very unlikely in this case unless the name of the member changes, in which case the header changes as well.
- Also it fosters uniformity and a better declaration/definition separation.

Bottom line: It very likely doesn't matter in this case from a runtime performance point of view, would have to be profiled in any case. It probably increases compile time performance very slightly (but also negligible b/c it's not a header that would be included in a lot of places or very high up the include hierarchy). Meh :)

Comment (clonker) [reply libyul/backends/evm/SSACFGTopologicalSort.cpp]: I've put them in the header now :laughing: the benefits from putting it into a cpp are in this case negligible anyways

Comment (matheusaaguiar) [reply libyul/backends/evm/SSACFGTopologicalSort.cpp]: I was more curious than anything here :)
I thought about the first two points you mentioned and was wondering if there were any other considerations.
Thanks for the explanation @clonker !

**Review by matheusaaguiar — APPROVED** [review]
